### PR TITLE
14803 add approval type component

### PR DESCRIPTION
**Pull request by jonathan-longe**
*Issue #:* /bcgov/entity14803

#### Default
image

#### Default with `isCourtOrderOnly` = true
image

#### Default with `isConversionToFullRestoration` = true
image

#### Draft with `approvedByRegistrar` = true
image

#### Draft with `draftCourtOrderNumber` = '99-1234567'
image

#### With Validation error shown
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:22]: I prefer to use the full syntax for the event handler, so it's clear that it's a method and not a property, ie, `emitValidationError()`.

The downside is that you have to explicitly pass the event data (`$event`), which is inherent in the first synatax.

Not a deal-breaker.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:64]: I counsel against identifying a specific filing type here... it would be better to use a flag named after the desired behaviour. (This component should not know what's going on in its parent.)

Also, comment is outdated.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:55]: I believe this is an object, not an enum. Seems to work though?

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:133]: If it's just a label changing, IMO you should pass the label string as a prop. It will keep this component more generic.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:148]: Unused CSS?

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:127]: This is fine, but FYI, another way this can be done is in the component itself, like this:
```
   :rules='enableValidation ? courtNumberRules : []
```

Where enableValidation is a flag you togle and courtNumberRules is a readonly const.

Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:127]: Do these rules ever need to be cleared?

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:43]: What are these 2 imports?

File: src/components/approval-type/package.json
Comment (severinbeauvais) [inline src/components/approval-type/package.json:11]: 👍

File: tests/unit/ApprovalType.spec.ts
Comment (severinbeauvais) [inline tests/unit/ApprovalType.spec.ts:33]: Not async.

Not a deal-breaker.

File: tests/unit/ApprovalType.spec.ts
Comment (severinbeauvais) [inline tests/unit/ApprovalType.spec.ts:67]: I'm not certain if it will work here, but sometimes you can simply await the setValue() above.

Sometimes you need to wait for the next timeslice (remember, JS is single-threaded) so that validation can take place and set the properties you expect to see changed.

**Review by severinbeauvais — COMMENTED** [review]
Looks really good. Clearly you got into the depths of everything :)

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:133]: Ah, good thought, I'll change it.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:22]: Good call, will fix

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:64]: Good call, I'll fix

File: tests/unit/ApprovalType.spec.ts
Comment (jonathan-longe) [reply tests/unit/ApprovalType.spec.ts:67]: Good to know.  If it's okay for you, I'll leave this as is.

File: tests/unit/ApprovalType.spec.ts
Comment (jonathan-longe) [reply tests/unit/ApprovalType.spec.ts:33]: Good catch.  I'll fix it

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:148]: Nice catch, removed

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:43]: I've removed what I think are the unnecessary import statements.  Let me know if you see any that are still not required.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:127]: Good to know.  I'll leave this as is for now, but I appreciate the feedback.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue]: Sometimes imports are automatically added by an IDE extension when you (accidentally) type in some things. I bet that's what these are.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:125]: Do these rules ever need to be cleared?

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue]: Thanks for update.

Same thing on line 8 actually.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:53]: I bet this would work as an enum (`enum ApprovalTypes = { etc`).

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue]: Looks good.

**Review by severinbeauvais — APPROVED** [review]
Please wait for another review before merging. I should not be the primary or only reviewer.

**Comment by jonathan-longe** [comment]
> Please wait for another review before merging. I should not be the primary or only reviewer.

Thanks @severinbeauvais !

File: src/components/approval-type/ApprovalType.stories.ts
Comment (cameron-eyds) [inline src/components/approval-type/ApprovalType.stories.ts:7]: Super nit picky, but probably doesn't need a new line. ie `argtypes: {}`

File: src/components/approval-type/ApprovalType.vue
Comment (cameron-eyds) [inline src/components/approval-type/ApprovalType.vue:24]: Once again, super nitpicky, we try to lead with the id, followed by properties/bound properties and subsequently event handlers. 
So in this case, maybe just move the `input` event to bottom with `update` etc. 
Not required at all, just general practice.

File: src/components/approval-type/ApprovalType.vue
Comment (cameron-eyds) [inline src/components/approval-type/ApprovalType.vue:30]: similar  to the above comment, leading with a conditional (when present) is what we've done in the past, with regards to arrangement of the component props. 
ie `<v-radio v-if="...`

File: src/components/approval-type/ApprovalType.vue
Comment (cameron-eyds) [inline src/components/approval-type/ApprovalType.vue:98]: types on the arguments might be helpful here.

File: src/components/approval-type/ApprovalType.vue
Comment (cameron-eyds) [inline src/components/approval-type/ApprovalType.vue:130]: No action required but an alternative to the string concatenation is use template strings. 
ie `This ${this.filingType} is approved by court order.` 
Just wrap in back ticks.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:98]: If it were me, I'd name the first one "valid" and the second one "value" (or val).

**Review by cameron-eyds — APPROVED** [review]
Looks good to me! 
Just some minor comments 👍

**Review by JazzarKarim — APPROVED** [review]
Looks good to me!

**Review by ketaki-deodhar — APPROVED** [review]
Lots of learning here for me :)

File: src/components/approval-type/ApprovalType.stories.ts
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.stories.ts:7]: Nice, I'll fix it!

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:24]: Thanks for letting me know.  I'll make the changes you suggest.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:30]: Makes sense. I'll fix it.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:98]: I agree, that's better.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:130]: Nice, that's prettier for sure.  I'll fix it.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:53]: prefixing ApprovalTypes with `enum` causes the unit tests to fail.  Are you okay if I leave this as is?

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:125]: No, not as far as I know.  Do I need a public clear() method?  Is that what you're suggesting?

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:53]: Yup.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:125]: I was thinking that you might need to set the rules to [] to disable them.

But maybe this is "once enabled, always enabled".

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:94]: We usually use the `@Emit` decorator instead of this, but it's valid.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [inline src/components/approval-type/ApprovalType.vue:111]: Why can't this be a simple assignment?

Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:111]: My understanding is that simple assignments are not reactive.

File: src/components/approval-type/ApprovalType.vue
Comment (jonathan-longe) [reply src/components/approval-type/ApprovalType.vue:94]: Yes, but the `@Emit` decorator doesn't allow me to change the value of the event.  On line 22, I call `@update:error="emitValidationError($event)"`  If there's a better way to do this, please let me know.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:111]: This PR is water under the bridge, but I'm fairly certain simple assignments _are_ reactive;  we use them all the time.

My understanding is that Vue.set adds a property to an existing reactive object (ie, related to deep reactivity). I think it's _probably_ not needed here.

File: src/components/approval-type/ApprovalType.vue
Comment (severinbeauvais) [reply src/components/approval-type/ApprovalType.vue:94]: If you return the value in an emit method, it should pass the value along, eg.

```
@Emit('myEvent')
protected emitMyEvent (myEvent: string): string {
   return myEvent
}
```

Or even simpler syntax:
```
@Emit('myEvent') protected emitMyEvent (myEvent: string): void {}
```

**Pull request merged**
